eliot/tests/test_action.py

```diff
@@ -1389,3 +1393,62 @@ class MyException(Exception):
                               "reason": "because",
                               "exception":
                               "eliot.tests.test_action.MyException"})
+
+
+
+class PreserveContextTests(TestCase):
+    """
+    Tests for L{preserve_context}.
+    """
+    def add(self, x, y):
+        """
+        Add two inputs.
+        """
+        Message.log(message_type="child")
+        return x + y
+
+    def test_no_context(self):
+        """
+        If C{preserve_context} is run outside an action context it just
+        returns the same function.
+        """
+        wrapped = preserve_context(self.add)
+        self.assertEqual(wrapped(2, 3), 5)
+
+    def test_with_context_calls_underlying(self):
+        """
+        If run inside an Eliot context, the result of C{preserve_context} is
+        the result of calling the underlying function.
+        """
+        with startAction(action_type="parent"):
+            wrapped = preserve_context(self.add)
+            self.assertEqual(wrapped(3, y=4), 7)
```

Comment:
Lines 1424 and 1425 in a helper function that this test method and `test_no_context` can both call?

:question: